src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs

```diff
@@ -114,26 +241,63 @@ private void Initialize(Compilation comp, AnalyzerExecutor analyzerExecutor, Can
             }
         }
 
-        private Task ExecuteSyntaxTreeActions(CancellationToken cancellationToken)
+        private async Task ExecuteSyntaxTreeActions(AnalysisScope analysisScope, AnalysisState analysisStateOpt, CancellationToken cancellationToken)
         {
+            if (analysisScope.IsTreeAnalysis && !analysisScope.IsSyntaxOnlyTreeAnalysis)
+            {
+                // For partial analysis, only execute syntax tree actions if performing syntax analysis.
+                return;
+            }
+
             // Execute syntax tree analyzers in parallel.
-            var tasks = ArrayBuilder<Task>.GetInstance();
-            foreach (var tree in _compilation.SyntaxTrees)
+            var tasks = analysisScope.ConcurrentAnalysis ? ArrayBuilder<Task>.GetInstance() : null;
+
+            try
             {
-                var actionsByAnalyzers = this.analyzerActions.SyntaxTreeActions.GroupBy(action => action.Analyzer);
-                foreach (var analyzerAndActions in actionsByAnalyzers)
+                foreach (var tree in analysisScope.SyntaxTrees)
                 {
-                    var task = Task.Run(() =>
+                    var actionsByAnalyzers = this.analyzerActions.SyntaxTreeActions.ToDictionary(a => a.Analyzer);
```

Comment:
good observation, switched it to create a map by analyzer like for other action kinds.